### PR TITLE
Move drawing and logic functionality into the BomberBlokeScene class. Added pickup functionality.

### PR DESCRIPTION
**Pull request by joeyshuttleworth**

### CODE REVIEW

File: common/BomberBlokeScene.cpp [hunk 1]
```diff
@@ -0,0 +1,212 @@
+#include "BomberBlokeScene.hpp"
```
Comment (tylerkf) [inline common/BomberBlokeScene.cpp:1]: Capital B on bloke???

File: engine/camera/Camera.cpp [hunk 2]
```diff
@@ -9,50 +9,57 @@ void Camera::rumble(double amplitude, double timeout){
...
     // TODO: Move this somewhere else
     // Set zoom value
-    double min = mWidth;
-    if (mHeight > min)
-        min = mHeight;
-    mZoom = ((double)min / mpScene->mDimmension[0]);
-    log_message(DEBUG, "Zoom scene is " + std::to_string(mZoom));
+    double min = width;
+    if (height > min)
+        min = height;
```
Comment (tylerkf) [inline engine/camera/Camera.cpp]: Maybe use fmin? more readable

File: engine/client.cpp [hunk 3]
```diff
@@ -24,11 +24,15 @@ void client_loop(){
      if(delay>0){
        SDL_Delay(delay);
      }
-     handle_input();
+     const std::lock_guard<std::mutex> lock(_scene_mutex);
```
Comment (tylerkf) [inline engine/client.cpp:28]: Please add a comment for this line

File: engine/engine.cpp [hunk 4]
```diff
@@ -269,20 +265,17 @@ SDL_Joystick *handle_input_controller() {
...
 
 void draw_screen() {
     if(_halt || !_renderer || !_window || !_draw)
         return;
-
-    SDL_SetRenderDrawColor(_renderer, 0x10, 0xFF, 0x00, 0xFF);
+    SDL_SetRenderDrawColor(_renderer, 0x00, 0x00, 0x00, 0xFF);
     SDL_RenderClear(_renderer);
-    _pCamera->draw();
+    _pScene->draw();
```
Comment (tylerkf) [inline engine/engine.cpp]: Is this safe? Shouldn't we check if there is a scene first?

File: engine/engine.hpp [hunk 5]
```diff
@@ -129,7 +133,6 @@ extern std::list<LocalPlayer> _local_player_list;
 const std::array<std::string, 2> _system_commands  = {{"bind", "set"}};
 
 class Camera;
```
Comment (tylerkf) [inline engine/engine.hpp]: Do we still need the Camera class here?

File: engine/hud/AbstractHudElement.cpp [hunk 6]
```diff
@@ -31,7 +31,7 @@ void AbstractHudElement::updatePosition(Camera* camera) {
     switch(mAlignFlags[0]) {
         case ALIGN_CENTER:
             // Centred positioning
-            mPosition[0] = (screenDimensions[0] - mDimensions[0]) / 2;
+          mPosition[0] = (screenDimensions[0] - mDimensions[0]) / 2;
```
Comment (tylerkf) [inline engine/hud/AbstractHudElement.cpp:34]: This is an accident

File: engine/hud/AbstractHudElement.hpp [hunk 7]
```diff
@@ -109,6 +109,39 @@ class AbstractHudElement {
...
+            case ALIGN_RIGHT:
+                // Right-aligned positioning
+                mPosition[1] = screenDimensions[1] - mDimensions[1] - mRelativePosition[1];
+                break;
+            default:
+                // Left-aligned positioning
+                mPosition[1] = mRelativePosition[1];
+        }
+    }
+
```
Comment (tylerkf) [inline engine/hud/AbstractHudElement.hpp:144]: No changes to this file should have been made

File: engine/hud/ClickableHudElement.cpp [hunk 8]
```diff
@@ -22,6 +22,10 @@ void ClickableHudElement::onInput(SDL_Event *event) {
     // If its a mouse button up event, it cannot be clicked
     if (mIsClicked && event->type == SDL_MOUSEBUTTONUP) {
         mIsClicked = false;
+        return;
+    }
+
+    if (event->type == SDL_MOUSEBUTTONDOWN) {
```
Comment (tylerkf) [inline engine/hud/ClickableHudElement.cpp:28]: This is an accidental change

File: common/BomberBlokeScene.hpp [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+    BomberBlokeScene(int size_x = 10, int size_y = 10);
+    ~BomberBlokeScene(){}
+
+   /**
+     * Every second, check how many players are alive.
+     * If no players remain, everybody draws. Otherwise,
+     * if one player remains, this player wins the match.
+     *
+     */
+    void LogicUpdate();
```
Comment (tylerkf) [inline common/BomberBlokeScene.hpp]: Is this meant to be camel case?

File: engine/scene/scene.hpp [hunk 10]
```diff
@@ -96,6 +118,13 @@ class scene{
   void movementUpdate();
   void physicsUpdate();
 
+  /**
+   *  A virtual function which can be overridden by child classes to implement
+   *  game specific logic.
+   */
+
+  virtual void LogicUpdate(){};
```
Comment (tylerkf) [inline engine/scene/scene.hpp]: Camel case?

File: engine/server.cpp [hunk 11]
```diff
@@ -26,12 +27,15 @@ void server_loop(){
     if(_tick % (5 * TICK_RATE) == 0){
       _ping_time = _tick;
     }
+    const std::lock_guard<std::mutex> lock(_scene_mutex);
```
Comment (tylerkf) [inline engine/server.cpp:31]: Please add a comment

File: server/bomberbloke_server.cpp [hunk 12]
```diff
@@ -21,83 +22,31 @@ int main (){
...
+
   server_loop();
   SDL_Quit();
   SDL_Delay(1000);
   return 0;
 }
 
 void new_game(std::string){
-  /* First set reset everyone's powerups */
+  const std::lock_guard<std::mutex> lock(_scene_mutex);
```
Comment (tylerkf) [inline server/bomberbloke_server.cpp:36]: Please add a comment

File: common/BomberBlokeScene.cpp [hunk 13]
```diff
@@ -0,0 +1,212 @@
...
+}
+
+static void hudTestFnJoin(){
+  handle_system_command({"open", "127.0.0.1"});
+}
+
+static void hudTestFn2() {
+  std::cout << "clicked left button" << std::endl;
+  handle_system_command(split_to_tokens("nickname dave1"));
+  handle_system_command(split_to_tokens("open 127.0.0.1"));
```
Comment (tylerkf) [inline common/BomberBlokeScene.cpp]: Remove this line, in default this line is not here

File: common/BomberBlokeScene.cpp [hunk 14]
```diff
@@ -0,0 +1,212 @@
...
+static void hudTestFn1() {
+  std::cout << "clicked centre button" << std::endl;
+}
+
+static void hudTestFnJoin(){
+  handle_system_command({"open", "127.0.0.1"});
+}
+
+static void hudTestFn2() {
+  std::cout << "clicked left button" << std::endl;
```
Comment (tylerkf) [inline common/BomberBlokeScene.cpp]: Remove this line, in default this line is not here

File: common/BomberBlokeScene.cpp [hunk 15]
```diff
@@ -0,0 +1,212 @@
...
+  handle_system_command(split_to_tokens("nickname dave1"));
+  handle_system_command(split_to_tokens("open 127.0.0.1"));
+}
+
+static void hudTestFn3() {
+  std::cout << "clicked right button" << std::endl;
```
Comment (tylerkf) [inline common/BomberBlokeScene.cpp]: Remove this line, in default this line is not here

File: common/BomberBlokeScene.cpp [hunk 16]
```diff
@@ -0,0 +1,212 @@
...
+  handle_system_command(split_to_tokens("nickname dave2"));
+  handle_system_command(split_to_tokens("open 127.0.0.1"));
```
Comment (tylerkf) [inline common/BomberBlokeScene.cpp]: Remove this line, in default this line is not here

File: common/BomberBlokeScene.cpp [hunk 17]
```diff
@@ -0,0 +1,212 @@
...
+#include "scene.hpp"
+#include "bomberbloke.h"
+#include "ClickableHudElement.hpp"
+#include "TextHudElement.hpp"
+#include "TextButton.hpp"
+#include "SpriteHudElement.hpp"
+
+static void hudTestFn1() {
+  std::cout << "clicked centre button" << std::endl;
+}
```
Comment (tylerkf) [inline common/BomberBlokeScene.cpp]: Remove this, in default this is not here

File: common/BomberBlokeScene.cpp [hunk 18]
```diff
@@ -0,0 +1,212 @@
...
+
+static void hudTestFnJoin(){
+  handle_system_command({"open", "127.0.0.1"});
+}
+
+static void hudTestFn2() {
```
Comment (tylerkf) [inline common/BomberBlokeScene.cpp:21]: Rename this to hudTestFn1 (there shouldn't be a third one)

File: common/BomberBlokeScene.cpp [hunk 19]
```diff
@@ -0,0 +1,212 @@
...
+  std::cout << "clicked left button" << std::endl;
+  handle_system_command(split_to_tokens("nickname dave1"));
+  handle_system_command(split_to_tokens("open 127.0.0.1"));
+}
+
+static void hudTestFn3() {
```
Comment (tylerkf) [inline common/BomberBlokeScene.cpp]: Rename this to hudTestFn2 (there shouldn't be a third one)

File: common/BomberBlokeScene.cpp [hunk 20]
```diff
@@ -0,0 +1,212 @@
...
+    mHudElements.push_back(hudElement);
+  }
+
+  // Power HUD demo
+  for (int i = 0; i < 3; i++) {
+    std::shared_ptr<SpriteHudElement> hudElement = std::make_shared<SpriteHudElement>("flames.png", 9 + i * 34, 9, 32, 32, ALIGN_RIGHT);
+    mHudElements.push_back(hudElement);
+  }
+
+  onResize();
```
Comment (tylerkf) [inline common/BomberBlokeScene.cpp]: Why do we call onResize on construction? This looks like an artifact from old default code

**Review by tylerkf — CHANGES_REQUESTED** [review]
Good work!!

**Comment by davenewham** [comment]
This resolves #17

**Pull request merged**